### DISCUSSION TITLE
Select the function to dump via fuzzy finder

### DISCUSSION THREAD

**Discussion by gitpy** [pacak/cargo-show-asm#195] [Ideas]
Is interactive selection via a fuzzy finder an interesting idea or should it be out of scope?

The simple solution would just check PATH for well known fuzzy finders.
For other fuzzy finders it could use a env variable first.

A more complex solution would add a fuzzy finder from library. IMO that's not worth the effort.

Another option would be to add something like `--list` for easy piping to fuzzy finders and an example of a shell alias to the documentation.

Proper preview support is only possible in the complex solution. It also works for the other solutions, but it might not scale too well on large code bases as it needs to rerun the command each time.

For the first solution I have written a PoC [here](https://github.com/gitpy/cargo-show-asm/commit/5c60a2f07b201a5f76ddbfcf9c0214243bb53cfb).

If there is interest for it, I might put something together.

**Comment by pacak** [comment]
This looks great, especially with the preview option! Rerunning does not scale well indeed, scanning 400Mb `.s` file takes a few seconds. I wonder if we can cheat by connecting to currently running `cargo-show-asm` process using loopback network of some sort. But even in current state it feels like an improvement.

External fuzzy finder seems like a way to go. Please let me know if you have any questions about the code :)

**Reply by gitpy** [reply]
To clarify with external fuzzy finder do you mean like the PoC or shell function/alias like I used in the past:
``` zsh
rasmp() {
  local build preview
  build=(cargo asm $@)

  preview="cargo asm --color ${@} {} 2> /dev/null"
  if [ $? -eq 0 ]; then
     cargo asm $@ 2> /dev/null | awk -F'"' 'NR>=2 {print $2}' \
     | fzf --ansi --preview-window up,60%,border-horizontal --preview "$preview"
  else
    echo $build;
  fi
}
```
Note: it's outdated and not entirely correct, because there is no unique status code for build successful but no single matching function

I will try making something with InterProcessCommunication, but it will take some time as I don't have a lot of experience with it on windows.

**Reply by pacak** [reply]
Something like you did in your PoC.

**Reply by pacak** [reply]
> but it will take some time as I don't have a lot of experience with it on windows.

No rush, not much experience with windows here either but networking should be the same I think.

**Comment by gitpy** [comment]
I have a [WIP](https://github.com/pacak/cargo-show-asm/compare/master...gitpy:cargo-show-asm:6b54ba21d9736b35555ebe27fc305d100da2743b) ready. I had to create a Trait over `dump_range` so it's a bit noisy. But the main thing is `indermediate_mode()` and `src/ipc.rs`

- I added the `interprocess` crate as a dependency. Not sure if that is maintained/used enough for you. Otherwise changing it to the network stack isn't much work.
- The whole preview and server+client is behind a feature flag
- I tried to hide the new `--client` arguments from the documentation but couldn't figure out how.
- The single thread server is fast enough from what I have tested

Let me know,  what you think.

**Reply by pacak** [reply]
> I added the interprocess crate as a dependency.

I'm not familiar with this crate, dependencies seem sane and the author is active. What do we gain by using it compared to just using the network sockets?

Will take a look at the code in a bit.

**Reply by gitpy** [reply]
It works over named pipes. So in the OS it skips the network stack. From my understanding it needs to make one copy less
Edit: Also skipping the network stack, potentially also skips weird firewall configurations.

**Reply by pacak** [reply]
https://www.baeldung.com/linux/ipc-performance-comparison - I don't think it will make any difference in performance as far as users are concerned, but yea, weird firewall might be a problem, let's stick with `interprocess` for now.

**Reply by pacak** [reply]
Changes look generally fine and switching works instant in a huge executable I tried earlier.

It would be great to add some tests for this feature specifically, probably not with `fzf` but with a custom stript. The rest - clean up TODOs and commented out code and add whatever else you have in mind. Maybe a few comments in the ipc module about the protocol.
